zwrap/sample.go

```diff
@@ -98,70 +103,62 @@ func (s *sampler) With(fields ...zap.Field) zap.Logger {
 }
 
 func (s *sampler) Check(lvl zap.Level, msg string) *zap.CheckedMessage {
-	if !s.check(lvl, msg) {
-		return nil
+	cm := s.Logger.Check(lvl, msg)
+	switch lvl {
+	case zap.PanicLevel, zap.FatalLevel:
+		return cm
 	}
-	return zap.NewCheckedMessage(s.Logger, lvl, msg)
+	if s.sampled(msg) {
+		return cm
+	}
+	return nil
 }
 
 func (s *sampler) Log(lvl zap.Level, msg string, fields ...zap.Field) {
-	if s.check(lvl, msg) {
+	if s.Logger.Check(lvl, msg) != nil && s.sampled(msg) {
 		s.Logger.Log(lvl, msg, fields...)
 	}
 }
 
 func (s *sampler) Debug(msg string, fields ...zap.Field) {
-	if s.check(zap.DebugLevel, msg) {
+	if s.Logger.Check(zap.DebugLevel, msg) != nil && s.sampled(msg) {
 		s.Logger.Debug(msg, fields...)
 	}
 }
 
 func (s *sampler) Info(msg string, fields ...zap.Field) {
-	if s.check(zap.InfoLevel, msg) {
+	if s.Logger.Check(zap.InfoLevel, msg) != nil && s.sampled(msg) {
 		s.Logger.Info(msg, fields...)
 	}
 }
 
 func (s *sampler) Warn(msg string, fields ...zap.Field) {
-	if s.check(zap.WarnLevel, msg) {
+	if s.Logger.Check(zap.WarnLevel, msg) != nil && s.sampled(msg) {
 		s.Logger.Warn(msg, fields...)
 	}
 }
 
 func (s *sampler) Error(msg string, fields ...zap.Field) {
-	if s.check(zap.ErrorLevel, msg) {
+	if s.Logger.Check(zap.ErrorLevel, msg) != nil && s.sampled(msg) {
 		s.Logger.Error(msg, fields...)
 	}
 }
 
 func (s *sampler) Panic(msg string, fields ...zap.Field) {
-	if s.check(zap.PanicLevel, msg) {
-		s.Logger.Panic(msg, fields...)
-	}
+	s.Logger.Panic(msg, fields...)
 }
 
 func (s *sampler) Fatal(msg string, fields ...zap.Field) {
-	if s.check(zap.FatalLevel, msg) {
-		s.Logger.Fatal(msg, fields...)
-	}
+	s.Logger.Fatal(msg, fields...)
 }
```

Comment:
Since we've embedded the unsampled logger, do we need `Panic` and `Fatal` on the sampler? We should get delegation for free.